File: materials/skills/hooks/codex-session-name-set.py
#!/usr/bin/env python3
"""Rename a Codex thread through the local shared app-server."""

import base64
import hashlib
import json
import os
import socket
import struct
import sys
from pathlib import Path


TIMEOUT_SECONDS = 2
WEBSOCKET_GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"


class BufferedSocket:
    def __init__(self, sock, initial=b""):
        self.sock = sock
        self.buffer = initial

    def read_exact(self, size):
        while len(self.buffer) < size:
            chunk = self.sock.recv(max(4096, size - len(self.buffer)))
            if not chunk:
                raise EOFError("app-server closed the socket")
            self.buffer += chunk
        data, self.buffer = self.buffer[:size], self.buffer[size:]
        return data


def send_frame(sock, opcode, payload):
    mask = os.urandom(4)
    length = len(payload)
    if length < 126:
        header = bytes([0x80 | opcode, 0x80 | length])
    elif length < 65536:
        header = bytes([0x80 | opcode, 0x80 | 126]) + struct.pack("!H", length)
    else:
        header = bytes([0x80 | opcode, 0x80 | 127]) + struct.pack("!Q", length)
    masked = bytes(byte ^ mask[index % 4] for index, byte in enumerate(payload))
    sock.sendall(header + mask + masked)


def send_json(sock, value):
    send_frame(sock, 0x1, json.dumps(value, separators=(",", ":")).encode())


def receive_json(reader):
    while True:
        first, second = reader.read_exact(2)
        if not first & 0x80:
            raise ValueError("fragmented app-server WebSocket frame")
        opcode = first & 0x0F
        length = second & 0x7F
        if length == 126:
            length = struct.unpack("!H", reader.read_exact(2))[0]
        elif length == 127:
            length = struct.unpack("!Q", reader.read_exact(8))[0]
        mask = reader.read_exact(4) if second & 0x80 else None
        payload = reader.read_exact(length)
        if mask:
            payload = bytes(
                byte ^ mask[index % 4] for index, byte in enumerate(payload)
            )
        if opcode == 0x1:
            return json.loads(payload)
        if opcode == 0x8:
            raise EOFError("app-server closed the WebSocket")
        if opcode == 0x9:
            send_frame(reader.sock, 0xA, payload)


def receive_response(reader, request_id):
    while True:
        message = receive_json(reader)
        if not isinstance(message, dict):
            raise ValueError("malformed app-server response")
        if message.get("id") != request_id:
            continue
        if "error" in message:
            raise RuntimeError(message["error"])
        if "result" not in message:
            raise ValueError("app-server response is missing result")
        return message["result"]


def connect(socket_path):
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    sock.settimeout(TIMEOUT_SECONDS)
    sock.connect(socket_path)

    key = base64.b64encode(os.urandom(16)).decode()
    sock.sendall(
        (
            "GET / HTTP/1.1\r\n"
            "Host: localhost\r\n"
            "Upgrade: websocket\r\n"
            "Connection: Upgrade\r\n"
            f"Sec-WebSocket-Key: {key}\r\n"
            "Sec-WebSocket-Version: 13\r\n\r\n"
        ).encode()
    )

    response = b""
    while b"\r\n\r\n" not in response:
        chunk = sock.recv(4096)
        if not chunk:
            raise EOFError("app-server closed during WebSocket handshake")
        response += chunk
    headers, initial = response.split(b"\r\n\r\n", 1)
    expected_accept = base64.b64encode(
        hashlib.sha1((key + WEBSOCKET_GUID).encode()).digest()
    ).decode()
    if (
        b"101 Switching Protocols" not in headers
        or expected_accept.encode() not in headers
    ):
        raise ConnectionError("app-server WebSocket handshake failed")
    return sock, BufferedSocket(sock, initial)


def set_thread_name(thread_id, name):
    codex_home = Path(os.environ.get("CODEX_HOME", Path.home() / ".codex"))
    socket_path = os.environ.get(
        "CODEX_APP_SERVER_SOCKET",
        str(codex_home / "app-server-control" / "app-server-control.sock"),
    )
    sock, reader = connect(socket_path)
    try:
        # Codex app-server protocol: initialize request, initialized notification,
        # then the stable thread/name/set request with {threadId, name}.
        send_json(
            sock,
            {
                "method": "initialize",
                "id": 1,
                "params": {
                    "clientInfo": {
                        "name": "codex-session-namer",
                        "title": "Codex Session Namer",
                        "version": "1.0.0",
                    }
                },
            },
        )
        receive_response(reader, 1)
        send_json(sock, {"method": "initialized", "params": {}})
        send_json(
            sock,
            {
                "method": "thread/name/set",
                "id": 2,
                "params": {"threadId": thread_id, "name": name},
            },
        )
        receive_response(reader, 2)
    finally:
        sock.close()


def main():
    if len(sys.argv) != 3 or not sys.argv[1] or not sys.argv[2]:
        return 2
    try:
        set_thread_name(sys.argv[1], sys.argv[2])
    except (OSError, EOFError, ValueError, RuntimeError):
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: test/codex-session-name-set.mjs
import assert from "node:assert/strict";
import { createHash } from "node:crypto";
import { spawn } from "node:child_process";
import { existsSync, mkdtempSync } from "node:fs";
import { createServer } from "node:net";
import { tmpdir } from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";

function ok(description) {
  console.log(`ok - ${description}`);
}

function runPython(helper, args, env = {}, timeoutMs = 3_000) {
  return new Promise((resolve) => {
    const child = spawn("python3", [helper, ...args], {
      env: { ...process.env, ...env },
      stdio: ["ignore", "pipe", "pipe"],
    });
    let stderr = "";
    let timedOut = false;
    const timer = setTimeout(() => {
      timedOut = true;
      child.kill("SIGKILL");
    }, timeoutMs);
    child.stderr.on("data", (chunk) => {
      stderr += chunk;
    });
    child.on("close", (status) => {
      clearTimeout(timer);
      resolve({ status, stderr, timedOut });
    });
  });
}

function websocketFrame(value) {
  const payload = Buffer.from(JSON.stringify(value));
  assert(payload.length < 126);
  return Buffer.concat([Buffer.from([0x81, payload.length]), payload]);
}

function readClientFrame(buffer) {
  if (buffer.length < 2) return null;
  let offset = 2;
  let length = buffer[1] & 0x7f;
  if (length === 126) {
    if (buffer.length < 4) return null;
    length = buffer.readUInt16BE(2);
    offset = 4;
  }
  assert.notEqual(length, 127, "測試訊息不該大到需要 64-bit frame length");
  const masked = (buffer[1] & 0x80) !== 0;
  assert.equal(masked, true, "WebSocket client frame 必須 mask");
  if (buffer.length < offset + 4 + length) return null;
  const mask = buffer.subarray(offset, offset + 4);
  offset += 4;
  const payload = Buffer.from(buffer.subarray(offset, offset + length));
  for (let index = 0; index < payload.length; index += 1) {
    payload[index] ^= mask[index % 4];
  }
  return {
    value: JSON.parse(payload.toString("utf8")),
    rest: buffer.subarray(offset + length),
  };
}

let appServerCase = 0;

async function runAppServerCase({ dir, responses, threadId = "thread-55" }) {
  appServerCase += 1;
  const socketPath = path.join(dir, `case-${appServerCase}.sock`);
  const messages = [];
  const server = createServer((socket) => {
    let buffer = Buffer.alloc(0);
    let upgraded = false;
    socket.on("data", (chunk) => {
      buffer = Buffer.concat([buffer, chunk]);
      if (!upgraded) {
        const end = buffer.indexOf("\r\n\r\n");
        if (end === -1) return;
        const headers = buffer.subarray(0, end).toString("utf8");
        const key = headers.match(/^Sec-WebSocket-Key: (.+)$/im)?.[1].trim();
        assert(key);
        const accept = createHash("sha1")
          .update(`${key}258EAFA5-E914-47DA-95CA-C5AB0DC85B11`)
          .digest("base64");
        socket.write(
          "HTTP/1.1 101 Switching Protocols\r\n" +
            "Upgrade: websocket\r\n" +
            "Connection: Upgrade\r\n" +
            `Sec-WebSocket-Accept: ${accept}\r\n\r\n`,
        );
        buffer = buffer.subarray(end + 4);
        upgraded = true;
      }

      while (true) {
        const frame = readClientFrame(buffer);
        if (frame === null) return;
        buffer = frame.rest;
        messages.push(frame.value);
        if (frame.value.id === 1) {
          socket.write(websocketFrame(responses.initialize));
        } else if (frame.value.id === 2) {
          socket.write(websocketFrame(responses.rename));
        }
      }
    });
  });
  await new Promise((resolve) => server.listen(socketPath, resolve));
  const result = await runPython(HELPER, [threadId, "修正原生命名"], {
    CODEX_APP_SERVER_SOCKET: socketPath,
  });
  await new Promise((resolve) => server.close(resolve));
  return { messages, result };
}

const REPO_ROOT = fileURLToPath(new URL("..", import.meta.url));
const HELPER = path.join(
  REPO_ROOT,
  "materials",
  "skills",
  "hooks",
  "codex-session-name-set.py",
);

try {
  assert.equal(existsSync(HELPER), true, "POSIX app-server helper 必須存在");

  const invalid = await runPython(HELPER, []);
  assert.equal(invalid.status, 2);
  ok("helper 參數錯誤時 exit 2");

  const unavailable = await runPython(HELPER, ["thread-a", "測試名稱"], {
    CODEX_APP_SERVER_SOCKET: path.join(tmpdir(), "missing-codex-app-server.sock"),
  });
  assert.equal(unavailable.status, 1);
  ok("helper 連不到 app-server 時 exit 1");

  const dir = mkdtempSync(path.join(tmpdir(), "jr-codex-app-server-"));
  const eofSocketPath = path.join(dir, "premature-eof.sock");
  const eofServer = createServer((socket) => socket.destroy());
  await new Promise((resolve) => eofServer.listen(eofSocketPath, resolve));
  const prematureEof = await runPython(
    HELPER,
    ["thread-eof", "不該卡住"],
    { CODEX_APP_SERVER_SOCKET: eofSocketPath },
    750,
  );
  eofServer.close();
  assert.equal(prematureEof.timedOut, false);
  assert.equal(prematureEof.status, 1);
  ok("WebSocket handshake 提前 EOF 時立即 exit 1，不等 socket timeout");

  const { messages, result } = await runAppServerCase({
    dir,
    responses: {
      initialize: { id: 1, result: {} },
      rename: { id: 2, result: {} },
    },
  });

  assert.equal(result.status, 0, result.stderr);
  assert.deepEqual(
    messages.map(({ method }) => method),
    ["initialize", "initialized", "thread/name/set"],
  );
  assert.deepEqual(messages[2].params, {
    threadId: "thread-55",
    name: "修正原生命名",
  });
  ok("helper 依序 initialize、initialized、thread/name/set 並成功 exit 0");

  const missingResult = await runAppServerCase({
    dir,
    responses: {
      initialize: { id: 1, result: {} },
      rename: { id: 2 },
    },
  });
  assert.equal(missingResult.result.status, 1);
  ok("matching id 缺少 result 時 exit 1");

  const appServerError = await runAppServerCase({
    dir,
    responses: {
      initialize: { id: 1, result: {} },
      rename: { id: 2, error: { code: -1, message: "rename failed" } },
    },
  });
  assert.equal(appServerError.result.status, 1);
  ok("matching id 回傳 error 時 exit 1");

  const malformed = await runAppServerCase({
    dir,
    responses: {
      initialize: [],
      rename: { id: 2, result: {} },
    },
  });
  assert.equal(malformed.result.status, 1);
  ok("matching response 不是 object 時 exit 1");
} catch (error) {
  console.error(`not ok - ${error.stack ?? error.message}`);
  process.exit(1);
}
